### PR TITLE
add bencode parser diferencial fuzzer

### PR DESCRIPTION
**Pull request by daphneherlambda**

### CODE REVIEW

**Comment by dafifynn** [comment]
to run the fuzzer use de command `cargo hfuzz run bencode-parser-fuzz -i ../../torrents`

this will run the fuzzer with the bencode files in torrents folder as a seed

**Comment by dafifynn** [comment]
to run criterion use cargo bench, at this point its going to use one of the seeds files to run, in further steps its going to use one of the files generated by the fuzzer run before

**Comment by daphneherlambda** [comment]
the code in nom-bencode/src/lib.rs fails in cases of error in
`60  	impl<I> From<nom::Err<Error<I>>> for Error<I> {`
`-> 61  	    fn from(e: nom::Err<Error<I>>) -> Self {`
`   62  	        e.into()`
`   63  	    }`
`   64  	}`
with the following error: 
Captura de pantalla 2023-01-11 a la(s) 12 58 04

so the fuzzer crashes at some errors that are well handled in libtorrent-rs

File: bencode-parser-fuzz/bencode-parser-fuzz/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+fn to_bencode(value: Vec<Value>) -> Bencode {
+    
+    match &value[..] {
+        [Value::Integer(value), _] => Bencode::BNumber(*value),
+        [Value::Bytes(value),_] => to_string_bencode(value),
+        [Value::List(value), _] => to_list_bencode(value),
+        [Value::Dictionary(value), _] => to_dict_bencode(value),
+        [] | [_] | [_, _, _, ..] => to_bencode(value),
```
Comment (Oppen) [inline bencode-parser-fuzz/bencode-parser-fuzz/src/main.rs]: You're passing `value` unmodified recursively to `to_bencode`, so I think this triggers infinite recursion. I don't entirely understand the last case in this line either. This should just be the terminating case I believe.

File: bencode-parser-fuzz/bencode-parser-fuzz/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+fn hashmap_to_btreemap(hashmap: &HashMap<&[u8], Value>) -> BTreeMap<Vec<u8>, Bencode> {
+    let mut btreemap: BTreeMap<Vec<u8>, Bencode> = BTreeMap::new();
+
+    btreemap = hashmap
+        .into_iter()
+        .map(|(key, value)| (from_reference_to_vec(key), value_to_bencode(value)))
+        .collect();
+    btreemap
+}
```
Comment (Oppen) [inline bencode-parser-fuzz/bencode-parser-fuzz/src/main.rs]: Missing newline.

File: bencode-parser-fuzz/bencode-parser-fuzz/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+fn hashmap_to_btreemap(hashmap: &HashMap<&[u8], Value>) -> BTreeMap<Vec<u8>, Bencode> {
+    let mut btreemap: BTreeMap<Vec<u8>, Bencode> = BTreeMap::new();
+
+    btreemap = hashmap
+        .into_iter()
+        .map(|(key, value)| (from_reference_to_vec(key), value_to_bencode(value)))
+        .collect();
+    btreemap
```
Comment (Oppen) [inline bencode-parser-fuzz/bencode-parser-fuzz/src/main.rs]:
```suggestion
    hashmap
        .into_iter()
        .map(|(key, value)| (from_reference_to_vec(key), value_to_bencode(value)))
        .collect();
```

File: bencode-parser-fuzz/bencode-parser-fuzz/src/main.rs (hunk 1 above)
Comment (daphneherlambda) [reply bencode-parser-fuzz/bencode-parser-fuzz/src/main.rs]: fixed! now it panic in that case, it has to be one of the previous values or else it makes no sense

File: bencode-parser-fuzz/bencode-parser-fuzz/src/main.rs (hunk 3 above)
Comment (daphneherlambda) [reply bencode-parser-fuzz/bencode-parser-fuzz/src/main.rs]: done

File: bencode-parser-fuzz/bencode-parser-fuzz/src/main.rs (hunk 2 above)
Comment (daphneherlambda) [reply bencode-parser-fuzz/bencode-parser-fuzz/src/main.rs]: done

**Comment by edg-l** [comment]
> the code in nom-bencode/src/lib.rs fails in cases of error in `60 impl<I> From<nom::Err<Error<I>>> for Error<I> {` `-> 61 fn from(e: nom::Err<Error<I>>) -> Self {` ` 62 e.into()` ` 63 }` ` 64 }` with the following error: Captura de pantalla 2023-01-11 a la(s) 12 58 04
> 
> so the fuzzer crashes at some errors that are well handled in libtorrent-rs

I updated the nom-bencode library, 0.2.0 should fix the issue.

**Comment by daphneherlambda** [comment]
this commit add the honggfuzz report of a new error in libtorrent-rs library, along with the triggering input, the error is the following:
Captura de pantalla 2023-02-08 a la(s) 11 22 39

**Review by igaray — CHANGES_REQUESTED** [review]
@daphneherlambda could you add your instructions for running the fuzz tests in a README file in the corresponding directory?

**Review by igaray — APPROVED** [review]

**Pull request merged**